src/audio/asrc/asrc_farrow.c

```diff
@@ -402,23 +403,25 @@ enum asrc_error_code asrc_initialise(struct comp_dev *dev,
 	}
 
 	/*
-	 * Set the pointer for the impulse response.
-	 * &m_impulse_response[0] = ring_buffers_x
-	 * + num_channels * (1 + buffer_length);
-	 * Here ring_buffer_x already points to the memory just behind
-	 * the src_obj.  The offset results from the number of pointers
-	 * pointing to each channel and each channels buffer data.
+	 * Set the pointer for the impulse response. It is just after
+	 * the ring_buffers32[] or ring_buffers16[] array.
 	 */
+
+	/* Error if odd buffer length */
+	if (src_obj->buffer_length & 1) {
+		comp_err(dev, "asrc_initialise(), odd buffer_length = %d",
+			 src_obj->buffer_length);
+		return ASRC_EC_INVALID_BUFFER_LENGTH;
+	}
+
 	if (src_obj->bit_depth == 32) {
 		src_obj->impulse_response =
-			(int32_t *)(src_obj->ring_buffers32 +
-				    src_obj->num_channels *
-				    (1 + src_obj->buffer_length));
+			&src_obj->ring_buffers32[src_obj->num_channels - 1]
+				[src_obj->buffer_length - 1] + 1;
 	} else if (src_obj->bit_depth == 16) {
-		src_obj->impulse_response =
-			(int32_t *)(src_obj->ring_buffers16 +
-				    src_obj->num_channels *
-				    (1 + src_obj->buffer_length / 2));
+		p16 = &src_obj->ring_buffers16[src_obj->num_channels - 1]
+				[src_obj->buffer_length  - 1] + 1;
+		src_obj->impulse_response = (int32_t *)p16;
```

Comment:
The standard way to deal with this sort of situation is to make `impulse_response` a `void *` and cast on the _reading_ side.

Reply:
I'd not prefer to do it since impulse_response[] is always 32 bit signed integer data and it's accessed in a lot of places in the code. The delay line for filter can be 16 or 32 bit and impulse_response[] location need to be computed differently that creates this extra code to save data RAM.

Reply:
> The delay line for filter can be 16 or 32 bit and impulse_response[] location need to be computed differently that creates this extra code to save data RAM.

Can you add this or something like it as a comment to explain what this 16-32 bits conversion is about?

In my ideal world every cast has a comment explaining the need to circumvent the type system. I realize embedded is very far from my ideal world but every little helps :-)

Reply:
Computing the offset in [][] way is hardly verifiable against the `asrc_get_required_size()` that attempts to compute size of the buffer that is split here. How about coding it consistently with the `asrc_get_required_size()` logic and use src_obj+1 as a byte array. Then the pointer arithmetic would be simple and readable.
```c
src_obj_bufs = (uint8_t*)(src_obj+1);
ring_buffers = (...)src_obj_bufs;
src_obj_bufs+= buffer_length * num_channels * (bit_depth / 8);
impulse_response = (...)src_obj_bufs;
```
No if-s, no [][], and formula is consistently coded.

Btw, there are two other potential issues with this cont. mem code:
- asrc_get_required_size() uses h/c 128, 256 max sizes for filter and buffer length, and they are not verified against values in c_filter_params[],
- there is no oob verification in this function (could be easily checked before assigning next ptr using ->asrc_size and src_obj_bufs value as an extension to the approach I proposed above).

Reply:
> Computing the offset in [][] way is hardly verifiable against the asrc_get_required_size() that attempts to compute size of the buffer that is split here.

I could be wrong but I think you're collapsing two different checks into a single one:
- the buffer(s) are big enough for the arrays
- the array indexes never go out of array bounds.

In general I find it more readable to have these two checks separate. I admit I don't have the time to understand this particular code and see how unusual it is.

The `[]` array syntax is strictly equivalent to pointer arithmetic, it's just a higher-level abstraction of the same thing. If counting bytes was more readable, arrays would not exist in C.

Reply:
I agree, that multi-dimensional array syntax is preferred where such arrays actually are located, but I think ASRC code is confusing how multi-dimensional arrays are implemented in C. A 2D array in C like `int array[3][2];` is a sequence of 6 integers `a[0][0], a[0][1], a[1][0]...` Whereas `asrc_get_required_size()` seems to suggest allocating an array of pointers to further arrays. Although I then don't see any loop that would assign addresses to that first array, so, it seems the buffer is then indeed handled as a 2D array. But at least the size calculation and possibly some other locations have to be cleaned up.

Reply:
I think this can be marked as resolved.